src/firefly/java/edu/caltech/ipac/firefly/server/visualize/fitseval/SpectralCubeEval.java

```diff
@@ -30,167 +24,143 @@
             params.put("id", "IpacTableFromSource");
             params.put("startIdx", "0");
             params.put("pageSize", "30000");
-            params.put("tbl_index", tableHdu.hduIdx+"");
+            params.put("tbl_index", tableHdu.hduIdx + "");
             params.put("source", ServerContext.replaceWithPrefix(f));
             relatedDataList.add(
-                    RelatedData.makeWavelengthTabularRelatedData(params, "WAVE-TAB", "Table Wavelength information", tableHdu.hduName, tableHdu.hduIdx));
+                    RelatedData.makeWavelengthTabularRelatedData(params, "WAVE-TAB", "Table Wavelength information",
+                            tableHdu.hduName, tableHdu.hduVersion, tableHdu.hduLevel, tableHdu.hduIdx));
 
         }
         return relatedDataList;
     }
 
 
     public static List<TableHdu> findWaveTabHDU(BasicHDU<?>[] HDUs) {
-        List<TableHdu> tableHduList= new ArrayList<>();
-        for(int i=0; (i<HDUs.length); i++) {
-            var altProjList= FitsReadUtil.getAtlProjectionIDs(HDUs[i].getHeader());
-            var tabHdu= findWaveTabHDU(HDUs,i,"");
-            if (tabHdu!=null) tableHduList.add(tabHdu);
-            for(String alt : altProjList) {
-                tabHdu= findWaveTabHDU(HDUs,i,alt);
-                if (tabHdu!=null) tableHduList.add(tabHdu);
+        List<TableHdu> tableHduList = new ArrayList<>();
+        for (int i = 0; (i < HDUs.length); i++) {
+            var altProjList = FitsReadUtil.getAtlProjectionIDs(HDUs[i].getHeader());
+            var tabHdu = findWaveTabHDU(HDUs, i, "");
+            if (tabHdu != null) tableHduList.add(tabHdu);
+            for (String alt : altProjList) {
+                tabHdu = findWaveTabHDU(HDUs, i, alt);
+                if (tabHdu != null) tableHduList.add(tabHdu);
             }
         }
         return tableHduList;
     }
 
-    private record TableHdu(int hduIdx, String hduName) {};
+    private record TableHdu(int hduIdx, String hduName, int hduVersion, int hduLevel) {}
 
+    private record ExtId(String extName, int extVer, int extLevel) {}
 
     /**
      * If a FITS file contains multiple XTENSION HDUs (headerdata
-     * units) with the specified EXTNAME, EXTLEVEL, and
-     * EXTVER, then the result of the WCS table lookup is undefined.
+     * units) with the specified EXTNAME, EXTVER, and EXTLEVEL,
+     * then the result of the WCS table lookup is undefined.
      * If the specified FITS BINTABLE contains no column, or multiple
      * columns, with the specified TTYPEn, then the result of the
      * WCS table lookup is undefined. The specified FITS BINTABLE
      * must contain only one row.
-     * No units conversions are to
-     * @return
+     *
+     * @param HDUs FITS HDUs
+     * @param tabExtId Extension ID of the table lookup extension
+     * @return matching tab extension null if it can not be found
      */
-    private static boolean isValidWCSForTab(BasicHDU<?>[] HDUs){
```

Comment:
so we don't need to validate at all?

Reply:
> so we don't need to validate at all?

`isValidWCSForTab` was checking that there are no extensions with binary tables that have the same EXTNAME, EXTVER, and EXTLEVEL. (For more than one matching extension, table lookup is undefined.) 

In the updated code, I check for this condition in `findTabExtension`: when looking for matching extension we need to do the same loop and WCS matching, no need to do it twice.

I'll update Javadoc to make it clear and fix the wrong brace placement in line 93.